### PR TITLE
Jc/python interface

### PR DESCRIPTION
**Pull request by joao-conde**
| - | - |
| --- | --- |
| Issue | (no github issue) Adds code to export our crate as a python module |
| Dependencies | [*chrono*](https://github.com/chronotope/chrono) as a **build dependency** only |
| Decisions | 1. Added `build.rs` file to generate static constants (such as `COMPILE_TIME`)  2. Provides export as a python module 3. Python module has constants (such as `VERSION` or `ALGORITHMS`), `blend_images` and `blend_multiple` functions (complete compatibility with currently used C library PConvert) |

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -1,11 +1,22 @@
 [package]
-name = "pconvert-rust"
+name = "pconvert_rust"
```
Comment (joamag) [inline Cargo.toml:2]: we really need to rename this to use the underscore?

File: build.rs [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+use std::io::Write;
+use std::path::Path;
+use std::process::Command;
+use std::str;
+
+const BUILD_OUT_FILE: &str = "constants.rs";
+const SOURCE_DIR: &str = "./src";
+
+/*
+Build script (https://doc.rust-lang.org/cargo/reference/build-scripts.html)
```
Comment (joamag) [inline build.rs:12]: Let's change the docstring according to this https://doc.rust-lang.org/stable/rust-by-example/meta/doc.html

File: src/pymodule.rs [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+                algorithm
+            ));
+            println!("Printing from Rust code loop with algorithm {}", algorithm);
+            let demultiply = is_algorithm_multiplied(&algorithm);
+            let algorithm_fn = get_blending_algorithm(&algorithm);
+            let current_layer = read_png(path, demultiply);
+            blend_images(&current_layer, &mut composition, &algorithm_fn);
+        }
+        write_png(out_path, &composition);
+        println!("Printing from Rust code #3");
```
Comment (joamag) [inline src/pymodule.rs:115]: remove println

File: src/pymodule.rs [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+        let mut composition = read_png(first_path, demultiply);
+        while let Some(pair) = zip_iter.next() {
+            let path = pair.0.unwrap().extract::<String>().unwrap();
+            let algorithm = pair.1;
+
+            let algorithm = BlendAlgorithm::from_str(algorithm).expect(&format!(
+                "Blending algorithm '{}' does not exist",
+                algorithm
+            ));
+            println!("Printing from Rust code loop with algorithm {}", algorithm);
```
Comment (joamag) [inline src/pymodule.rs:108]: remove println

File: src/pymodule.rs [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    ) {
+        println!("Printing from Rust code #1");
+        let num_images = img_paths.len().unwrap() as usize;
+
+        if num_images < 1 {
+            eprintln!("ERROR: Specify at least one image path");
+            std::process::exit(-1);
+        }
+
+        //TODO: actually make use of this
```
Comment (joamag) [inline src/pymodule.rs:70]: remove todo

File: src/pymodule.rs [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+
+    #[pyfn(m, "blend_multiple")]
+    fn blend_multiple_py(
+        img_paths: &PySequence,
+        out_path: String,
+        algorithm: Option<String>,
+        algorithms: Option<Vec<String>>,
+        is_inline: Option<bool>,
+    ) {
+        println!("Printing from Rust code #1");
```
Comment (joamag) [inline src/pymodule.rs:62]: remove println

**Review by joamag — CHANGES_REQUESTED** [review]

File: Cargo.toml (hunk 1 above)
Comment (joao-conde) [reply Cargo.toml:2]: Changed it to follow the [Rust naming conventions](https://doc.rust-lang.org/1.0.0/style/style/naming/README.html).
Furthermore, the `package` can be named `pconvert-rust` but the `lib` can not. So, for coherence and in order to follow the naming conventions, I suggest naming both `pconvert_rust`

**Review by joamag — APPROVED** [review]

**Review by joamag — APPROVED** [review]

**Pull request merged**